### PR TITLE
Prevent changing profiles if settings are not saved

### PR DESCRIPTION
**Pull request by haslinghuis**
Keep `isDirty` state consistent with firmware sliders.

Due to firmware sliders we save the current slider state to get PIDs etc loosing state in firmware until saved. So if user does not want to save we are storing and retrieving original state at tab load or using refresh.

Button `Reset all profiles` resets the `isDirty` flag so we can change to another profile while settings may have changed in the previous profile which could load saved settings to the wrong profile.

- set `isDirty` after reset profile button
- reset flags and resulting state after using refresh button.
- reset `isDirty` when changes are undone.

Problems solved:

- use `isDirty` and `sliderRetainConfiguration` flag to keep state and prevent going to another profile without saving.
- Fixes: #2619 reworked handlers and selectors (rates tab was using same selectors)
- disable input on RP or RPY according to `sliderMode`. This will users remind to save settings after changing `sliderMode`.
- Enabling PID sliders respects initial firmware setting (user can use reset profile button for setting sliders to default).

### CODE REVIEW

**Comment by haslinghuis** [comment]
The alternative - WIP

```
diff --git a/src/js/TuningSliders.js b/src/js/TuningSliders.js
index 09023f25..8e4dd607 100644
--- a/src/js/TuningSliders.js
+++ b/src/js/TuningSliders.js
@@ -32,7 +32,7 @@ const TuningSliders = {
     PID_DEFAULT: [],
     FILTER_DEFAULT: {},
     SLIDER_DEFAULT: {},
-    initialSettings: {},
+    initialSettings: [],
 
     cachedPidSliderValues: false,
     cachedGyroSliderValues: false,
@@ -43,57 +43,70 @@ const TuningSliders = {
 
 const D_MIN_RATIO = 0.85;
 
-TuningSliders.saveInitialSettings = function () {
+TuningSliders.saveInitialSettings = function (profile) {
     if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
-        this.initialSettings.sliderPidsMode = FC.TUNING_SLIDERS.slider_pids_mode;
-        this.initialSettings.sliderDGain = FC.TUNING_SLIDERS.slider_d_gain / 100;
-        this.initialSettings.sliderPIGain = FC.TUNING_SLIDERS.slider_pi_gain / 100;
-        this.initialSettings.sliderFeedforwardGain = FC.TUNING_SLIDERS.slider_feedforward_gain / 100;
-        this.initialSettings.sliderDMaxGain = FC.TUNING_SLIDERS.slider_dmax_gain / 100;
-        this.initialSettings.sliderIGain = FC.TUNING_SLIDERS.slider_i_gain / 100;
-        this.initialSettings.sliderRollPitchRatio = FC.TUNING_SLIDERS.slider_roll_pitch_ratio / 100;
-        this.initialSettings.sliderPitchPIGain = FC.TUNING_SLIDERS.slider_pitch_pi_gain / 100;
-        this.initialSettings.sliderMasterMultiplier = FC.TUNING_SLIDERS.slider_master_multiplier / 100;
-        this.initialSettings.sliderGyroFilter = FC.TUNING_SLIDERS.slider_gyro_filter;
-        this.initialSettings.sliderGyroFilterMultiplier = FC.TUNING_SLIDERS.slider_gyro_filter_multiplier / 100;
-        this.initialSettings.sliderDTermFilter = FC.TUNING_SLIDERS.slider_dterm_filter;
-        this.initialSettings.sliderDTermFilterMultiplier = FC.TUNING_SLIDERS.slider_dterm_filter_multiplier / 100;
+        this.initialSettings[profile] = {
+            'sliderPidsMode': FC.TUNING_SLIDERS.slider_pids_mode,
+            'sliderDGain': FC.TUNING_SLIDERS.slider_d_gain / 100,
+            'sliderPIGain': FC.TUNING_SLIDERS.slider_pi_gain / 100,
+            'sliderFeedforwardGain': FC.TUNING_SLIDERS.slider_feedforward_gain / 100,
+            'sliderDMaxGain': FC.TUNING_SLIDERS.slider_dmax_gain / 100,
+            'sliderIGain': FC.TUNING_SLIDERS.slider_i_gain / 100,
+            'sliderRollPitchRatio': FC.TUNING_SLIDERS.slider_roll_pitch_ratio / 100,
+            'sliderPitchPIGain': FC.TUNING_SLIDERS.slider_pitch_pi_gain / 100,
+            'sliderMasterMultiplier': FC.TUNING_SLIDERS.slider_master_multiplier / 100,
+            'sliderGyroFilter': FC.TUNING_SLIDERS.slider_gyro_filter,
+            'sliderGyroFilterMultiplier': FC.TUNING_SLIDERS.slider_gyro_filter_multiplier / 100,
+            'sliderDTermFilter': FC.TUNING_SLIDERS.slider_dterm_filter,
+            'sliderDTermFilterMultiplier': FC.TUNING_SLIDERS.slider_dterm_filter_multiplier / 100,
+        };
     }
 };
 
-TuningSliders.restoreInitialSettings = function () {
+TuningSliders.restoreInitialSettings = function (currentProfile) {
     if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
-        if (this.sliderModeHasChanged && this.initialSetting.sliderPidsMode !== this.sliderPidsmode) {
-            $('#sliderPidsModeSelect').val(this.initialSettings.sliderPidsMode).trigger('change');
-        }
-
-        FC.TUNING_SLIDERS.slider_pids_mode = this.initialSettings.sliderPidsMode;
-
-        FC.TUNING_SLIDERS.slider_d_gain = Math.round(this.initialSettings.sliderDGain * 20) * 5;
-        FC.TUNING_SLIDERS.slider_pi_gain = Math.round(this.initialSettings.sliderPIGain * 20) * 5;
-        FC.TUNING_SLIDERS.slider_feedforward_gain = Math.round(this.initialSettings.sliderFeedforwardGain * 20) * 5;
-        FC.TUNING_SLIDERS.slider_dmax_gain = Math.round(this.initialSettings.sliderDMaxGain * 20) * 5;
-        FC.TUNING_SLIDERS.slider_i_gain = Math.round(this.initialSettings.sliderIGain * 20) * 5;
-        FC.TUNING_SLIDERS.slider_roll_pitch_ratio = Math.round(this.initialSettings.sliderRollPitchRatio * 20) * 5;
-        FC.TUNING_SLIDERS.slider_pitch_pi_gain = Math.round(this.initialSettings.sliderPitchPIGain * 20) * 5;
-        FC.TUNING_SLIDERS.slider_master_multiplier = Math.round(this.initialSettings.sliderMasterMultiplier * 20) * 5;
+        console.log(this.initialSettings);
 
-        FC.TUNING_SLIDERS.slider_gyro_filter = this.initialSettings.sliderGyroFilter;
-        FC.TUNING_SLIDERS.slider_gyro_filter_multiplier = this.initialSettings.sliderGyroFilterMultiplier * 100;
-        FC.TUNING_SLIDERS.slider_dterm_filter = this.initialSettings.sliderDTermFilter;
-        FC.TUNING_SLIDERS.slider_dterm_filter_multiplier = this.initialSettings.sliderDTermFilterMultiplier * 100;
+        if (TABS.pid_tuning.sliderRetainMode[currentProfile] && this.initialSetting[currentProfile].sliderPidsMode !== this.sliderPidsmode) {
+            $('#sliderPidsModeSelect').val(this.initialSettings[currentProfile].sliderPidsMode).trigger('change');
+        }
 
-        MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS))
-        .then(() => MSP.promise(MSPCodes.MSP_PID))
-        .then(() => MSP.promise(MSPCodes.MSP_PID_ADVANCED))
-        .then(() => MSP.promise(MSPCodes.MSP_FILTER_CONFIG))
-        .then(() => {
-            TABS.pid_tuning.configChanges = {};
-            if (GUI.active_tab === 'pid_tuning') {
-                this.updateFormPids();
-                TABS.pid_tuning.updatePIDColors();
-            }
-        });
+        console.log(this.initialSettings.length);
+        for (let profile = 0; profile < this.initialSettings.length; profile++) {
+            MSP.promise(MSPCodes.MSP_SELECT_SETTING, profile).then(function () {
+                console.log('SELECTED PROFILE TO RESTORE SETTNGS', profile);
+                console.log(this.initialSettings[profile]);
+                FC.CONFIG.profile = profile;
+                FC.TUNING_SLIDERS.slider_pids_mode = this.initialSettings[profile].sliderPidsMode;
+
+                FC.TUNING_SLIDERS.slider_d_gain = Math.round(this.initialSettings[profile].sliderDGain * 20) * 5;
+                FC.TUNING_SLIDERS.slider_pi_gain = Math.round(this.initialSettings[profile].sliderPIGain * 20) * 5;
+                FC.TUNING_SLIDERS.slider_feedforward_gain = Math.round(this.initialSettings[profile].sliderFeedforwardGain * 20) * 5;
+                FC.TUNING_SLIDERS.slider_dmax_gain = Math.round(this.initialSettings[profile].sliderDMaxGain * 20) * 5;
+                FC.TUNING_SLIDERS.slider_i_gain = Math.round(this.initialSettings[profile].sliderIGain * 20) * 5;
+                FC.TUNING_SLIDERS.slider_roll_pitch_ratio = Math.round(this.initialSettings[profile].sliderRollPitchRatio * 20) * 5;
+                FC.TUNING_SLIDERS.slider_pitch_pi_gain = Math.round(this.initialSettings[profile].sliderPitchPIGain * 20) * 5;
+                FC.TUNING_SLIDERS.slider_master_multiplier = Math.round(this.initialSettings[profile].sliderMasterMultiplier * 20) * 5;
+
+                FC.TUNING_SLIDERS.slider_gyro_filter = this.initialSettings[profile].sliderGyroFilter;
+                FC.TUNING_SLIDERS.slider_gyro_filter_multiplier = this.initialSettings[profile].sliderGyroFilterMultiplier * 100;
+                FC.TUNING_SLIDERS.slider_dterm_filter = this.initialSettings[profile].sliderDTermFilter;
+                FC.TUNING_SLIDERS.slider_dterm_filter_multiplier = this.initialSettings[profile].sliderDTermFilterMultiplier * 100;
+
+                MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS))
+                .then(() => MSP.promise(MSPCodes.MSP_PID))
+                .then(() => MSP.promise(MSPCodes.MSP_PID_ADVANCED))
+                .then(() => MSP.promise(MSPCodes.MSP_FILTER_CONFIG))
+                .then(() => {
+                    TABS.pid_tuning.configChanges = {};
+                    if (GUI.active_tab === 'pid_tuning') {
+                        this.updateFormPids();
+                        TABS.pid_tuning.updatePIDColors();
+                    }
+                });
+            });
+        }
+        FC.CONFIG.profile = currentProfile;
     }
 };
 
diff --git a/src/js/main.js b/src/js/main.js
index 839c1f4b..a9153985 100644
--- a/src/js/main.js
+++ b/src/js/main.js
@@ -252,8 +252,8 @@ function startProcess() {
             const tabName = $(self).text();
 
             if (GUI.active_tab === 'pid_tuning') {
-                if (TABS.pid_tuning.sliderRetainPosition || TABS.pid_tuning.sliderRetainMode || TABS.pid_tuning.sliderRetainConfiguration) {
-                    TuningSliders.restoreInitialSettings();
+                if (TABS.pid_tuning.sliderRetainPosition[FC.CONFIG.profile] || TABS.pid_tuning.sliderRetainMode[FC.CONFIG.profile] || TABS.pid_tuning.sliderRetainConfiguration[FC.CONFIG.profile]) {
+                    TuningSliders.restoreInitialSettings(FC.CONFIG.profile);
                 }
             }
 
diff --git a/src/js/tabs/pid_tuning.js b/src/js/tabs/pid_tuning.js
index de32bc89..85d7c3a5 100644
--- a/src/js/tabs/pid_tuning.js
+++ b/src/js/tabs/pid_tuning.js
@@ -22,10 +22,10 @@ TABS.pid_tuning = {
     activeSubtab: 'pid',
     analyticsChanges: {},
 
-    sliderChanges: {},
-    sliderRetainMode: false,
-    sliderRetainPosition: false,
-    sliderRetainConfiguration: false,
+    sliderChanges: [],
+    sliderRetainMode: [ false, false, false ],
+    sliderRetainPosition: [ false, false, false ],
+    sliderRetainConfiguration: [ false, false, false ],
 };
 
 TABS.pid_tuning.initialize = function (callback) {
@@ -1369,7 +1369,7 @@ TABS.pid_tuning.initialize = function (callback) {
 
         $('#resetProfile').on('click', function(){
             self.updating = true;
-            self.sliderRetainConfiguration = true;
+            self.sliderRetainConfiguration[self.currentProfile] = true;
 
             MSP.promise(MSPCodes.MSP_SET_RESET_CURR_PID).then(function () {
                 self.refresh(function () {
@@ -1902,26 +1902,36 @@ TABS.pid_tuning.initialize = function (callback) {
             const SLIDER_STEP_UPPER = semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44) ? 0.05 : 0.1;
 
             if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
-                if (!self.sliderRetainConfiguration) {
-                    TuningSliders.saveInitialSettings();
+                if (!self.sliderRetainConfiguration[self.currentProfile]) {
+                    TuningSliders.saveInitialSettings(self.currentProfile);
+                } else {
+                    // TODO: remove when we restore all changed profiles
+                    self.setDirty(true);
                 }
 
-                const initialConfiguration = TuningSliders.initialSettings;
+                console.log('TUNINGSLIDERS INIT');
+                const initialConfiguration = TuningSliders.initialSettings[self.currentProfile];
 
                 // we only target the range target type.
                 function sliderHandler(e) {
                     if (e.target !== e.currentTarget) {
                         const item = e.target.id === '' ? e.target.name : e.target.id;
                         const value = isInt(e.target.value) ? parseInt(e.target.value) : parseFloat(e.target.value);
-                        self.sliderChanges[item] = value;
+                        self.sliderChanges[self.currentProfile] = { item: value };
+
+                        console.log(typeof self.sliderChanges);
+                        console.log(self.sliderChanges);
 
                         if (item in initialConfiguration) {
                             if (value !== initialConfiguration[item]) {
-                                self.sliderRetainPosition = true;
+                                self.sliderRetainPosition[self.currentProfile] = true;
                             } else {
-                                delete self.sliderChanges[item];
-                                if (Object.keys(self.sliderChanges).length === 0) {
-                                    self.sliderRetainPosition = false;
+                                console.log(Object.keys(self.sliderChanges[self.currentProfile].length));
+                                delete self.sliderChanges[self.currentProfile].item;
+                                console.log(Object.keys(self.sliderChanges[self.currentProfile].length));
+                                if (Object.keys(self.sliderChanges[self.currentProfile]).length === 0) {
+                                    console.log('Changes undone');
+                                    self.sliderRetainPosition[self.currentProfile] = false;
                                 }
                             }
                         }
@@ -1937,7 +1947,7 @@ TABS.pid_tuning.initialize = function (callback) {
                             e.preventDefault();
                         } else {
                             sliderPidsModeSelectElement.val(0).trigger('change');
-                            self.sliderRetainMode = true;
+                            self.sliderRetainMode[self.currentProfile] = true;
                         }
                     } else {
                         self.updateGuiElements();
@@ -2112,7 +2122,7 @@ TABS.pid_tuning.initialize = function (callback) {
                 //set Slider PID mode to RPY when re-enabling Sliders
                 if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
                     $('#sliderPidsModeSelect').val(2).trigger('change');
-                    self.sliderRetainMode = true;
+                    self.sliderRetainMode[self.currentProfile] = true;
                 }
                 // if values were previously changed manually and then sliders are reactivated, reset pids to previous valid values if available, else default
                 TuningSliders.resetPidSliders();
@@ -2190,7 +2200,7 @@ TABS.pid_tuning.initialize = function (callback) {
             });
             // enable Filter sliders button
             $('a.buttonFilterTuningSliders').click(function() {
-                self.sliderRetainMode = true;
+                self.sliderRetainMode[self.currentProfile] = true;
                 if (TuningSliders.GyroSliderUnavailable) {
                     //set Slider mode to ON when re-enabling Sliders
                     if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
@@ -2297,10 +2307,10 @@ TABS.pid_tuning.initialize = function (callback) {
             }).then(function () {
                 self.updating = false;
 
-                self.sliderRetainPosition = false;
-                self.sliderRetainMode = false;
-                self.sliderChanges = {};
-                self.sliderRetainConfiguration = false;
+                self.sliderRetainPosition[self.currentProfile] = false;
+                self.sliderRetainMode[self.currentProfile] = false;
+                self.sliderChanges[self.currentProfile] = {};
+                self.sliderRetainConfiguration[self.currentProfile] = false;
 
                 self.setDirty(false);
 
@@ -2411,15 +2421,13 @@ TABS.pid_tuning.cleanup = function (callback) {
 TABS.pid_tuning.refresh = function (callback) {
     const self = this;
 
-    if ((self.sliderRetainPosition || self.sliderRetainMode) && !self.updating) {
-        TuningSliders.restoreInitialSettings();
+    if ((self.sliderRetainPosition[self.currentProfile] || self.sliderRetainMode[self.currentProfile]) && !self.updating) {
+        TuningSliders.restoreInitialSettings(self.currentProfile);
     }
 
     GUI.tab_switch_cleanup(function () {
         self.initialize();
 
-        self.setDirty(false);
-
         if (callback) {
             callback();
         }
```

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by ctzsnooze — APPROVED** [review]

**Comment by ctzsnooze** [comment]
This works really well!  Tests fine on Mac OSX.

The `Reset all Profile Values` button now doesn't auto-save the reset... which is MUCH better behaviour.  The user can now reset to defaults to 'check out' what those defaults would be, without losing their existing settings.  They must, like any other change, hit the `SAVE` button if they want to save their changes.  

This is much more consistent behaviour and a definite improvement.

**Review by mikeller — APPROVED** [review]
Tested, seems to be working fine.

**Comment by mituritsyn** [comment]
Hi guys, seems this pr has some troubles. PID and some rates fields have became inactive. 
image
image

**Comment by haslinghuis** [comment]
@mituritsyn - please set RPY to OFF and save to be able to edit PID values manual.

Somehow Rates Roll Center and Expo are also effected (so have to check how this is happening as we use classes for selection).

**Pull request merged**
